api/authenticators/okta/authenticator.go

```diff
@@ -0,0 +1,166 @@
+	req := authnRequest{Username: creds.Username, Password: creds.Password}
+	res, err := a.oktaAuthClient.Authn(ctx, req)
+	// We don't need to acknowledge this error because we're using zero values all the way down
+	return core.User{ID: res.UserID()}, err
+}
+
+// ListApplications should list all the applications the given user is entitled to access.
+func (a *Authenticator) ListApplications(ctx context.Context, user core.User) ([]core.Application, error) {
+	// TODO: It seems like a bad idea to put this filtering here as it inherently makes it an AWS-okta provider, rather than just an Okta one.
+	// We use the app links endpoint because it's the easiest way to find the applications a user may access.
+	links, resp, err := a.client.User.ListAppLinks(ctx, user.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	for resp.HasNextPage() {
+		var next []*okta.AppLink
+		if resp, err = resp.Next(a.ctx, &next); err != nil {
+			return nil, err
+		}
+
+		links = append(links, next...)
+	}
+
+	var awsAccounts []core.Application
+	for _, app := range links {
+		if app.AppName != "amazon_aws" {
+			continue
+		}
+
+		awsAccounts = append(awsAccounts, core.Application{
+			LegacyID: 0,
+			ID:       app.AppInstanceId,
+			Name:     app.Label,
+		})
+	}
+
+	return awsAccounts, nil
+}
+
+type extractedRole struct {
+	AWSAccountName string
+	AWSAccountID   string
+	AWSRoleName    string
+	OktaGroupID    string
+}
+
+// extractAWSAccountName attempts to extract the account name and ID from the given Okta group.
+func extractRole(group *okta.Group) (extractedRole, bool) {
+	// TODO: Filtering the format this way in this location seems like a bad idea.
+	// It's certainly not trivial for other people to use.
+	var r extractedRole
+
+	// RG-AWS.account_name.role_name.account_id
+	split := strings.Split(group.Profile.Name, ".")
+	if split[0] != "RG-AWS" || len(split) != 4 {
+		return r, false
+	}
+
+	r.AWSAccountName = split[1]
+	r.AWSRoleName = split[2]
+	r.AWSAccountID = split[3]
+	r.OktaGroupID = group.Id
+	return r, true
+}
+
+// GenerateSAMLAssertion should generate a SAML assertion that the user may exchange with the target application in order to gain access to it.
+// This will initiate a multi-factor request with Duo.
+func (a *Authenticator) GenerateSAMLAssertion(ctx context.Context, creds core.Credentials, appID string) (*core.SAMLResponse, error) {
+	app, _, err := a.client.Application.GetApplication(ctx, appID, &okta.Application{}, query.NewQueryParams())
+	if err != nil {
+		return nil, err
+	}
+
+	appl := app.(*okta.Application)
+
+	st, err := a.oktaAuthClient.Authn(ctx, authnRequest{Username: creds.Username, Password: creds.Password})
+	if err != nil {
+		return nil, err
+	}
+
+	var f *okta.UserFactor
+	for _, factor := range st.Factors() {
+		if factor.Provider == "DUO" && factor.FactorType == "web" {
+			f = &factor
+			break
+		}
+	}
+
+	if f == nil {
+		// TODO: No factors that matched criteria - return sentinel error here
```

Comment:
This will fail if someone does not have a Duo web factor set up. A better implementation would be to return an actual meaningful error here, but at the time I took the bet that this was so rare that it would not happen. Users will see a blank error message if this occurs.

Reply:
Yeah, I've added this in the bug-fixes branch, but I'm going to cherry-pick that on top of this one.